internal/scheduler/reports.go

```diff
@@ -271,41 +271,129 @@ func extractQueueAndJobContexts(sctx *schedulercontext.SchedulingContext) (map[s
+		for executorId, qctx := range byExecutor {
+			if existing, existed := mostRecent[executorId]; existed && qctx.Created.Before(existing.Created) {
+				continue
+			}
+			if _, preempted := qctx.EvictedJobsById[jobId]; preempted {
+				mostRecentPreempting[executorId] = qctx
+			}
+		}
+	}
+
+	return schedulingReport{
+		mostRecentSchedulingContextByExecutor:           armadamaps.MapValues(mostRecent, schedulercontext.GetSchedulingContextFromQueueSchedulingContext),
+		mostRecentSuccessfulSchedulingContextByExecutor: armadamaps.MapValues(mostRecentSuccessful, schedulercontext.GetSchedulingContextFromQueueSchedulingContext),
+		mostRecentPreemptingSchedulingContextByExecutor: armadamaps.MapValues(mostRecentPreempting, schedulercontext.GetSchedulingContextFromQueueSchedulingContext),
+
+		sortedExecutorIds: repo.GetSortedExecutorIds(),
+	}
+}
+
+func (repo *SchedulingContextRepository) getSchedulingReport() schedulingReport {
+	return schedulingReport{
+		mostRecentSchedulingContextByExecutor:           repo.GetMostRecentSchedulingContextByExecutor(),
+		mostRecentSuccessfulSchedulingContextByExecutor: repo.GetMostRecentSuccessfulSchedulingContextByExecutor(),
+		mostRecentPreemptingSchedulingContextByExecutor: repo.GetMostRecentPreemptingSchedulingContextByExecutor(),
+
+		sortedExecutorIds: repo.GetSortedExecutorIds(),
+	}
+}
+
 // GetSchedulingReport is a gRPC endpoint for querying scheduler reports.
 // TODO: Further separate this from internal contexts.
-func (repo *SchedulingContextRepository) GetSchedulingReport(_ context.Context, _ *types.Empty) (*schedulerobjects.SchedulingReport, error) {
-	return &schedulerobjects.SchedulingReport{
-		Report: repo.getSchedulingReportString(),
-	}, nil
+func (repo *SchedulingContextRepository) GetSchedulingReport(_ context.Context, request *schedulerobjects.SchedulingReportRequest) (*schedulerobjects.SchedulingReport, error) {
+	var sr schedulingReport
+
+	switch filter := request.GetFilter().(type) {
+	case *schedulerobjects.SchedulingReportRequest_MostRecentForQueue:
+		queueName := strings.TrimSpace(filter.MostRecentForQueue.GetQueueName())
+		sr = repo.getSchedulingReportForQueue(queueName)
+	case *schedulerobjects.SchedulingReportRequest_MostRecentForJob:
+		jobId := strings.TrimSpace(filter.MostRecentForJob.GetJobId())
+		sr = repo.getSchedulingReportForJob(jobId)
+	default:
+		sr = repo.getSchedulingReport()
+	}
+
+	return &schedulerobjects.SchedulingReport{Report: sr.ReportString(request.GetVerbosity())}, nil
 }
 
-func (repo *SchedulingContextRepository) getSchedulingReportString() string {
-	sortedExecutorIds := repo.GetSortedExecutorIds()
-	mostRecentSchedulingContextByExecutor := repo.GetMostRecentSchedulingContextByExecutor()
-	mostRecentSuccessfulSchedulingContextByExecutor := repo.GetMostRecentSuccessfulSchedulingContextByExecutor()
-	mostRecentPreemptingSchedulingContextByExecutor := repo.GetMostRecentPreemptingSchedulingContextByExecutor()
+type schedulingReport struct {
```

Comment:
This struct exists so that it's slightly harder to confuse the different scheduling contexts when turning them into a string.